modules/navier_stokes/src/variables/BernoulliPressureVariable.C

```diff
@@ -0,0 +1,144 @@
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "BernoulliPressureVariable.h"
+#include "NS.h"
+#include "NSFVUtils.h"
+
+registerMooseObject("NavierStokesApp", BernoulliPressureVariable);
+
+InputParameters
+BernoulliPressureVariable::validParams()
+{
+  auto params = INSFVPressureVariable::validParams();
+  params += FunctorInterface::validParams();
+  params.addRequiredParam<MooseFunctorName>("u", "The x-component of velocity");
+  params.addParam<MooseFunctorName>("v", 0, "The y-component of velocity");
+  params.addParam<MooseFunctorName>("w", 0, "The z-component of velocity");
+  params.addRequiredParam<MooseFunctorName>(NS::porosity, "The porosity");
+  params.addRequiredParam<MooseFunctorName>(NS::density, "The density");
+  return params;
+}
+
+BernoulliPressureVariable::BernoulliPressureVariable(const InputParameters & params)
+  : INSFVPressureVariable(params),
+    FunctorInterface(this),
+    _u(nullptr),
+    _v(nullptr),
+    _w(nullptr),
+    _eps(nullptr),
+    _rho(nullptr)
+{
+}
+
+void
+BernoulliPressureVariable::initialSetup()
+{
+  _u = &getFunctor<ADReal>("u", _subproblem);
+  _v = &getFunctor<ADReal>("v", _subproblem);
+  _w = &getFunctor<ADReal>("w", _subproblem);
+  _eps = &getFunctor<ADReal>(NS::porosity, _subproblem);
+  _rho = &getFunctor<ADReal>(NS::density, _subproblem);
+}
+
+std::pair<bool, ADRealVectorValue>
+BernoulliPressureVariable::elemIsUpwind(const Elem & elem, const FaceInfo & fi) const
+{
+  const Moose::FaceArg face{&fi, Moose::FV::LimiterType::CentralDifference, true, false, nullptr};
+
+  const VectorValue<ADReal> vel_face{(*_u)(face), (*_v)(face), (*_w)(face)};
+  const bool fi_elem_is_upwind = vel_face * fi.normal() > 0;
+  const bool elem_is_upwind = &elem == &fi.elem() ? fi_elem_is_upwind : !fi_elem_is_upwind;
+
+  return {elem_is_upwind, vel_face};
+}
+
+bool
+BernoulliPressureVariable::isExtrapolatedBoundaryFace(const FaceInfo & fi,
+                                                      const Elem * const elem) const
+{
+  if (isDirichletBoundaryFace(fi, elem))
+    return false;
+  if (!isInternalFace(fi))
+    // We are neither a Dirichlet nor an internal face
+    return true;
+
+  // If we got here, then we're definitely on an internal face
+
+  if (std::get<0>(_eps->isDiscontinuous(fi)))
+    // We choose to extrapolate for the element that is downwind
+    return !std::get<0>(elemIsUpwind(*elem, fi));
+
+  return false;
+}
+
+bool
+BernoulliPressureVariable::isDirichletBoundaryFace(const FaceInfo & fi,
+                                                   const Elem * const elem) const
+{
+  if (INSFVPressureVariable::isDirichletBoundaryFace(fi, elem))
+    return true;
+
+  if (isInternalFace(fi) && std::get<0>(_eps->isDiscontinuous(fi)))
+    // We choose to apply the Dirichlet condition for the upwind element
+    return std::get<0>(elemIsUpwind(*elem, fi));
```

Comment:
Yeah, I think we should split these kind of functions into a check and an evaluate function.

Reply:
nah, you evaluate in order to perform the check. Don't want to waste. I'd be willing to rename the method though